service/src/event/validator.rs

```diff
@@ -0,0 +1,269 @@
+        match value {
+            DeliverableRequirement::Immediate => HistoryRequirement::Complete,
+            DeliverableRequirement::Asap => HistoryRequirement::InitOnly,
+            DeliverableRequirement::Lazy => HistoryRequirement::None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    history_req: HistoryRequirement,
+}
+
+impl<'a> EventValidator<'a> {
+    pub fn new(pool: &'a SqlitePool, history_req: HistoryRequirement) -> Self {
+        Self { pool, history_req }
+    }
+
+    pub(crate) async fn validate_events(
+        &self,
+        parsed_events: Vec<EventInsertable>,
+        invalid_events: &mut Vec<InvalidItem>,
+    ) -> Result<(Vec<EventInsertable>, Vec<EventInsertable>)> {
+        // first collect all the init events so we can refer to them when grouping by type
+        // in order to avoid querying the database or rejecting the events incorrectly.
+        // we will then add any missing init events from the DB for the next phase.
+        // guess at a capacity and just assume half are init events so we only need to re-allocate once at most
+        // TODO: we should validate these init events before we trust them for others.. could
+        // somehow validate data as okay but init as invalid (???)
+        let mut required_init_events =
+            std::collections::HashMap::with_capacity(parsed_events.len() / 2);
+        parsed_events.iter().for_each(|event| {
+            if event.event().is_init() {
+                required_init_events.insert(*event.cid(), event.event().to_owned());
+            }
+        });
+        // Group events by their type (over allocate vectors for now)
+        let mut valid_events = Vec::with_capacity(parsed_events.len());
+        let mut signed_events = Vec::with_capacity(parsed_events.len());
+        let mut time_events = Vec::with_capacity(parsed_events.len());
+        let mut pending_events = match self.history_req {
+            HistoryRequirement::InitOnly => Vec::with_capacity(parsed_events.len() / 2),
+            _ => Vec::new(),
+        };
+        for event in parsed_events {
+            match event.event().as_ref() {
+                unvalidated::Event::Time(time) => {
+                    let init_known = if required_init_events.contains_key(time.id()) {
+                        true
+                    } else {
+                        let (exists, _) =
+                            CeramicOneEvent::deliverable_by_cid(self.pool, time.id()).await?;
+                        exists
+                    };
+                    if init_known || matches!(self.history_req, HistoryRequirement::None) {
+                        time_events.push(event);
+                    } else if matches!(self.history_req, HistoryRequirement::InitOnly) {
+                        pending_events.push(event);
+                    } else {
+                        invalid_events.push(InvalidItem::RequiresHistory {
```

Comment:
this feels like more of a IOD concern than a validation concern.  I feel like all the validation code shouldn't need to know anything about the HistoryRequirement frankly.  From validation's perspective it either has everything it needs to do the validation, or it doesn't and is still missing something.  But whether the fact that missing something is an error or not is a question for a higher layer, not something the validation system should be concerning itself with.

Reply:
Maybe validation should return 3 vectors?  valid, invalid, and pending?  Then the caller can decide if it wants to turn pending into invalid.

Reply:
Yeah, I need to encapsulate this better and think about it a bit more. Right now it does return 3 vectors (sort of): `valid`, `pending` return by value and it access the mutable reference to invalid to add to it. I'm basically trying to avoid doing signature validation until we care/need to. If I can make a type that encapsulates the phases of validation, it'll be store the partially validated event and pick up where we need to.

The reason this exists is pretty much because of the migration. We can't validate migrated data events completely without the init events, but we're currently not willing to wait for them because it's already so time consuming and IO intensive and we get random ordering, so we want to validate signatures etc without checking things against the init, so at least we have some confidence the data was made by js-ceramic and is valid.